fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -0,0 +1,356 @@
+/*
+ * LucenePartitioner.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2023 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene;
+
+import com.apple.foundationdb.KeyValue;
+import com.apple.foundationdb.ReadTransaction;
+import com.apple.foundationdb.StreamingMode;
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.async.AsyncIterable;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.provider.foundationdb.IndexMaintainerState;
+import com.apple.foundationdb.tuple.Tuple;
+import com.google.common.base.Preconditions;
+import com.google.protobuf.InvalidProtocolBufferException;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static com.apple.foundationdb.record.provider.foundationdb.FDBStoreTimer.Waits.WAIT_LOAD_LUCENE_PARTITION_METADATA;
+
+/**
+ * Manage partitioning info for a <b>logical</b>, partitioned lucene index, in which each partition is a separate physical lucene index.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class LucenePartitioner {
+
+    public static final int PARTITION_META_SUBSPACE = 0;
+    public static final int PARTITION_DATA_SUBSPACE = 1;
+    private final IndexMaintainerState state;
+    private final boolean partitioningEnabled;
+    private final String partitionTimestampFieldName;
+    private final Map<Tuple, List<LucenePartitionInfoProto.LucenePartitionInfo>> partitionMetaInfoCache = new HashMap<>();
+
+    public LucenePartitioner(@Nonnull IndexMaintainerState state) {
+        this.state = state;
+        partitionTimestampFieldName = state.index.getOption(com.apple.foundationdb.record.metadata.IndexOptions.TEXT_DOCUMENT_PARTITION_TIMESTAMP);
```

Comment:
Note: the record layer has a concept of index validation that can be run to ensure that the options and expressions make sense. At some point we should add validation that field exists, and is of the right type.